crmprtd/bc_hydro/download.py

```diff
@@ -0,0 +1,116 @@
+''' Downloads data from BC hyrdo
+
+BC Hydro posts a rolling window (3 months) observing hourly data
+once a week. Data is in txt files.
+
+It is recommended to run this script once every 3 months to collect
+all available data, supplemented with aggressive monitoring and
+alerting for errors. If the script is run less than 3 months you
+will miss data.
+'''
+
+import pysftp
+from tempfile import TemporaryDirectory
+import logging
+import os
+import sys
+import re
+from zipfile import ZipFile
+from argparse import ArgumentParser
+from datetime import date, timedelta
+
+from crmprtd import logging_args, setup_logging, common_auth_arguments
+
+log = logging.getLogger(__name__)
+
+
+def download(username, gpg_private_key, ftp_server, ftp_dir, start_date,
+             end_date):
+
+    # Connect FTP server and retrieve file
+    try:
+        sftp = pysftp.Connection(ftp_server, username=username,
+                                 private_key=gpg_private_key)
+    except Exception:
+        log.exception("Invalid ftp authentication")
+        sys.exit(1)
+
+    try:
+        with TemporaryDirectory() as tmp_dir:
+            os.chdir(os.path.expanduser(tmp_dir))
+            range = DateRange(int(start_date), int(end_date), sftp)
+            sftp.walktree(ftp_dir, range.matchFile, matchDir, unknownType)
+            for filename in os.listdir(os.getcwd()):
+                name, extension = os.path.splitext(filename)
+                if extension == '.zip':
+                    zip_file = ZipFile(filename, 'r')
+                    for name in zip_file.namelist():
+                        txt_file = zip_file.open(name)
+                        sys.stdout.buffer.write(txt_file.read())
+
+    except IOError:
+        log.exception("Unable to download or open files")
+        sys.exit(1)
+
+
+class DateRange():
+
+    def __init__(self, start_date, end_date, sftp):
+        self.start_date = start_date
+        self.end_date = end_date
+        self.sftp = sftp
+
+    # Add files within date range to temp dir
+    def matchFile(self, filename):
+        match = re.search(r'[0-9]{6}', filename)
+        if match:
+            date = int(match.group(0))
+        if date >= self.start_date and date <= self.end_date:
+            self.sftp.get(filename)
+
+
+def matchDir(dir):
```

Comment:
If this (and `unknownType`) are currently unimplemented and unused, it's best to leave them out of the commit.

Reply:
These are because connection.walktree expects [4 arguments](https://pysftp.readthedocs.io/en/release_0.2.8/pysftp.html?highlight=walktree#pysftp.Connection.walktree), 3 of which are function calls. I got an error when I did not include them. Is there something I could do instead?

Reply:
Ah OK, I see why you did that.You're working around their API which is maybe has some... interesting design choices. Arguably they should have had no-op defaults for those arguments.

So yes, what you did is totally functional and works great. Stylistically, I'd argue for something *slightly* more succinct.

1. We can use a lambda expression instead of a function definition and we can use the same for both arguments. I.e.

```python3
no_op = lambda x: None
sftp.walktree(ftp_dir, range.matchFile, no_op, no_op)
```

2. I'd define `no_op` within the `download()` function, so that it retains local scope and isn't visible outside of the module (or even the function). That's a slightly picky point, but it helps keep the module namespace nice and clean.